gnovm/cmd/gno/repl.go

```diff
@@ -89,170 +85,93 @@ func execRepl(cfg *replCfg, args []string) error {
-	r := repl{
-		i:         1,
-		stdout:    stdout,
-		stderr:    stderr,
-		imports:   make([]string, 0),
-		funcs:     make([]string, 0),
-		lastInput: "// your code will be here", // initial value, to make it easier to identify with '/src'
-	}
-	for _, imp := range strings.Split(cfg.initialImports, ",") {
-		if strings.TrimSpace(imp) == "" {
-			continue
+// handleInput will read the input string and parse it depending if it
+// is a specific command, or source code. It returns true if the following
+// input is expected to be on more than one line.
+func handleInput(r *repl.Repl, input string) bool {
+	switch strings.TrimSpace(input) {
+	case "/reset":
+		r.Reset()
+	case "/src":
+		fmt.Fprintln(os.Stdout, r.Src())
+	case "/exit":
+		os.Exit(0)
+	case "/editor":
+		fmt.Fprintln(os.Stdout, "// Entering editor mode (^D to finish)")
+		return true
+	case "":
+		// avoid to increase the repl execution counter if sending empty content
+		return false
+	default:
+		out, err := r.Process(input)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
 		}
-		r.imports = append(r.imports, `import "`+imp+`"`)
-	}
-	testStore := tests.TestStore(cfg.rootDir, "", stdin, stdout, stderr, tests.ImportModeStdlibsOnly)
-	if cfg.verbose {
-		testStore.SetLogStoreOps(true)
-	}
-	r.machine = gno.NewMachineWithOptions(gno.MachineOptions{
-		PkgPath: "test",
-		Output:  stdout,
-		Store:   testStore,
-	})
-	defer r.machine.Release()
-
-	if cfg.initialCommand != "" {
-		r.handleInput(cfg.initialCommand)
+		fmt.Fprintln(os.Stdout, out)
 	}
 
-	// main loop
-	isTerm := term.IsTerminal(int(stdin.Fd()))
+	return false
+}
 
-	if isTerm {
-		rw := struct {
-			io.Reader
-			io.Writer
-		}{os.Stdin, os.Stderr}
-		t := term.NewTerminal(rw, "")
-		for {
-			// prompt and parse
-			t.SetPrompt(fmt.Sprintf("gno:%d> ", r.i))
-			oldState, err := term.MakeRaw(0)
-			if err != nil {
-				return fmt.Errorf("make term raw: %w", err)
-			}
-			input, err := t.ReadLine()
-			if err != nil {
-				term.Restore(0, oldState)
-				if errors.Is(err, io.EOF) {
-					return nil
-				}
-				return fmt.Errorf("term error: %w", err)
-			}
-			term.Restore(0, oldState)
+const (
+	inputBreaker = "^D"
+	nl           = "\n"
+)
 
-			err = r.handleInput(input)
-			if err != nil {
-				return fmt.Errorf("handle repl input: %w", err)
-			}
+func getInput(ml bool) (string, error) {
+	s := bufio.NewScanner(os.Stdin)
+	var mlOut bytes.Buffer
+	for s.Scan() {
+		line := s.Text()
+		if !ml {
+			return line, nil
 		}
-	} else { // !isTerm
-		scanner := bufio.NewScanner(stdin)
-		for scanner.Scan() {
-			input := scanner.Text()
-			err := r.handleInput(input)
-			if err != nil {
-				return fmt.Errorf("handle repl input: %w", err)
-			}
-		}
-		err := scanner.Err()
-		if err != nil {
-			return fmt.Errorf("read stdin: %w", err)
+
+		if line == inputBreaker {
+			break
 		}
+
+		mlOut.WriteString(line)
+		mlOut.WriteString(nl)
 	}
-	return nil
+
+	if err := s.Err(); err != nil {
+		return "", err
+	}
+
```

Comment:
Perhaps a minor, mostly aesthetic detail rather than anything significant, is handling the Ctrl+D command outside the `/editor` to either exit or simply jump a line
```suggestion
	// handle Ctrl+D editor
	if mlOut.Len() == 0 {
		if !ml {
			return "", io.EOF // exit (or at last skip a line?)
		}
		fmt.Print(nl) // newline
	}
```

Reply:
Fixed adding a new line if the processed input is empty:

```go
case "":
		// avoid to increase the repl execution counter if sending empty content
		fmt.Fprintln(os.Stdout, "")
		return false
```